src/EditorFeatures/Core/LanguageServer/AlwaysActiveLanguageClientEventListener.cs

```diff
@@ -40,8 +40,25 @@ internal sealed class AlwaysActiveLanguageClientEventListener(
     /// </summary>
     public void StartListening(Workspace workspace)
     {
-        // Trigger a fire and forget request to the VS LSP client to load our ILanguageClient.
-        Load();
+        _ = workspace.RegisterWorkspaceChangedHandler(Workspace_WorkspaceChanged);
+    }
+
+    private void Workspace_WorkspaceChanged(WorkspaceChangeEventArgs e)
+    {
+        if (e.Kind == WorkspaceChangeKind.SolutionAdded)
+        {
+            // Normally VS will load the language client when an editor window is created for one of our content types,
+            // but we want to load it as soon as a solution is loaded so workspace diagnostics work, and so 3rd parties
+            // like Razor can use dynamic registration.
+            Load();
+        }
+        else if (e.Kind is WorkspaceChangeKind.SolutionRemoved)
+        {
+            // VS will unload the language client when the solution is closed, but sometimes its a little slow to do so,
+            // and we can end up trying to load it, above, before it has been asked to unload. We wait for the previous
+            // instance to shutdown when we load, but we have to ensure that it at least gets signaled to do so first.
+            Unload();
```

Comment:
nit, calling these LoadSolution and UnloadSolution would be nice.

Reply:
sure!  i generally meant "what is unloading".  but up to you on picking new names :)